resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosCoarseGrainedSchedulerBackend.scala

```diff
@@ -393,7 +409,30 @@ private[spark] class MesosCoarseGrainedSchedulerBackend(
         val offerId = offer.getId.getValue
         val resources = remainingResources(offerId)
 
-        if (canLaunchTask(slaveId, resources)) {
+        var createTask = canLaunchTask(slaveId, resources)
+        if (hostToLocalTaskCount.nonEmpty) {
```

Comment:
It sounds like you're assuming `hostToLocalTaskCount.empty -> there are no executors to launch`.  That may be true, but regardless, if that's the case, then `createTask` is already false, so this check is redundant, right?

To make this more clear, please factor out this code into some function like `satisfiesLocality`, and instead of calling it here, for consistency, please add it as one of the constraints in `canLaunchTask`.